plenario/api/validator.py

```diff
@@ -0,0 +1,173 @@
+from collections import namedtuple
+from datetime import datetime, timedelta
+from dateutil import parser
+from marshmallow import fields, Schema
+from marshmallow.validate import Range, Length, OneOf
+
+from plenario.api.common import extract_first_geometry_fragment, make_fragment_str
+from plenario.database import session
+from plenario.models import MetaTable, ShapeMetadata
+
+
+class Validator(Schema):
+    """Base validator object using Marshmallow. Don't be intimidated! As scary
+    as the following block of code looks it's quite simple, and saves us from
+    writing validators. Let's break it down...
+
+    <FIELD_NAME> = fields.<TYPE>(default=<DEFAULT_VALUE>, validate=<VALIDATOR FN>)
+
+    The validator, when instanciated, has a method called 'dump'.which expects a
+    dictionary of arguments, where keys correspond to <FIELD_NAME>. The validator
+    has a default <TYPE> checker, that along with extra <VALIDATOR FN>s will
+    accept or reject the value associated with the key. If the value is missing
+    or rejected, the validator will substitue it with the value specified by
+    <DEFAULT_VALUE>."""
+
+    valid_aggs = {'day', 'week', 'month', 'quarter', 'year'}
+    valid_formats = {'csv', 'geojson', 'json'}
+
+    agg = fields.Str(default='week', validate=OneOf(valid_aggs))
+    buffer = fields.Integer(default=100, validate=Range(0))
+    dataset_name = fields.Str(default=None, validate=OneOf(MetaTable.index()), dump_to='dataset')
+    dataset_name__in = fields.List(fields.Str(), default=MetaTable.index(), validate=Length(1))
+    date__time_of_day_ge = fields.Integer(default=0, validate=Range(0, 23))
+    date__time_of_day_le = fields.Integer(default=23, validate=Range(0, 23))
+    data_type = fields.Str(default='json', validate=OneOf(valid_formats))
+    location_geom__within = fields.Str(default=None, dump_to='geom')
+    obs_date__ge = fields.Date(default=datetime.now() - timedelta(days=90))
+    obs_date__le = fields.Date(default=datetime.now())
+    offset = fields.Integer(default=0, validate=Range(0))
+    resolution = fields.Integer(default=500, validate=Range(0))
+    shape = fields.Str(default=None, validate=OneOf(ShapeMetadata.index()))
+
+
+class DatasetRequiredValidator(Validator):
```

Comment:
So, I'm a little uncomfortable with only having 3 Validator configs. Both `/detail-aggregate` and `/fields` use DatasetRequiredValidator, but `location_geom_within` makes no sense for `/fields`. So we can't send a warning back when a user adds a parameter for an endpoint that doesn't use that parameter. I think I'm OK with that tradeoff though, given the clarity of the code here.